### ISSUE TITLE
Test fails with Swift 5.5 on linux, but only when run in parallel or by itself: testSchedulingTaskOnSleepingLoopWakesUpOnce

### ISSUE THREAD

**Issue by buttaface** [apple/swift-nio#1971]
### Expected behavior
All tests pass when run in parallel with Swift 5.5.

### Actual behavior
One test consistently fails:
```
Test Case 'SALEventLoopTests.testSchedulingTaskOnSleepingLoopWakesUpOnce' started at 2021-09-27 19:34:37.105
Unblocking wedged task
<EXPR>:0: error: SALEventLoopTests.testSchedulingTaskOnSleepingLoopWakesUpOnce : threw error "TimeoutError(description: "/home/butta/swift-nio/Tests/NIOPosixTests/SyscallAbstractionLayer.swift:460")"
```
### Steps to reproduce

1. `../swift-5.5-RELEASE-centos8/usr/bin/swift test --parallel`
or
`../swift-5.5-RELEASE-centos8/usr/bin/swift test --filter OnSleepingLoop`

I cannot reproduce when all the tests are run serially, ie with `swift test` alone.

### SwiftNIO version/commit hash

latest HEAD, 06cb3dd

### Swift & OS version (output of `swift --version && uname -a`)

Swift 5.5 running on Fedora Core 33

**Comment by Lukasa** [comment]
Hmm, this doesn't immediately reproduce on my Ubuntu Swift 5.5 install. My suspicion here is that this is a timing-sensitive test, which would partially explain why you only see it under these circumstances. This timeout will fire if it takes more than 5 seconds to run the test. That implies there is an ordering issue.

**Comment by Lukasa** [comment]
Can I ask for more diagnostic details here? For example, can we get an entire call stack at the point of the error? You can do this by changing the `try` on line 468 to a `try!` and then using lldb.

**Comment by buttaface** [comment]
Here is the full demangled backtrace from within the lldb that comes with the Swift 5.5 toolchain:
```
(lldb) bt
* thread #2, name = 'NIO-ELT-0-#0', stop reason = Assertion failed: illegal state: runningAndAcceptingNewRegistrations
  * frame #0: 0x00007ffff7de3b70 libswiftCore.so`_swift_runtime_on_report
    frame #1: 0x00007ffff7e5c04f libswiftCore.so`_swift_stdlib_reportFatalErrorInFile + 207
    frame #2: 0x00007ffff7b69662 libswiftCore.so`closure #1 (Swift.UnsafeBufferPointer<Swift.UInt8>) -> () in closure #1 (Swift.UnsafeBufferPointer<Swift.UInt8>) -> () in closure #1 (Swift.UnsafeBufferPointer<Swift.UInt8>) -> () in Swift._assertionFailure(_: Swift.StaticString, _: Swift.String, file: Swift.StaticString, line: Swift.UInt, flags: Swift.UInt32) -> Swift.Never + 226
    frame #3: 0x00007ffff7b6938b libswiftCore.so`closure #1 (Swift.UnsafeBufferPointer<Swift.UInt8>) -> () in closure #1 (Swift.UnsafeBufferPointer<Swift.UInt8>) -> () in Swift._assertionFailure(_: Swift.StaticString, _: Swift.String, file: Swift.StaticString, line: Swift.UInt, flags: Swift.UInt32) -> Swift.Never + 107
    frame #4: 0x00007ffff7b6822f libswiftCore.so`Swift._assertionFailure(_: Swift.StaticString, _: Swift.String, file: Swift.StaticString, line: Swift.UInt, flags: Swift.UInt32) -> Swift.Never + 447
    frame #5: 0x00005555559d2f84 swift-nioPackageTests.xctest`$defer #1 (self=0x00007fffec001930) in SelectableEventLoop.run() at SelectableEventLoop.swift:429:13
    frame #6: 0x00005555559d26f4 swift-nioPackageTests.xctest`SelectableEventLoop.run(self=0x00007fffec001930) at SelectableEventLoop.swift:404:15
    frame #7: 0x000055555599c5e1 swift-nioPackageTests.xctest`static MultiThreadedEventLoopGroup.runTheLoop(thread=0x00007fffec000b60, canEventLoopBeShutdownIndividually=false, selectorFactory=0x0000555555ef3320 swift-nioPackageTests.xctest`partial apply forwarder for closure #3 () throws -> NIOPosix.Selector<NIOPosix.NIORegistration> in NIOPosixTests.SALTest.setUpSAL() -> () at <compiler-generated>, initializer=0x00005555559a1d80 swift-nioPackageTests.xctest`partial apply forwarder for reabstraction thunk helper from @escaping @callee_guaranteed (@in_guaranteed NIOPosix.NIOThread) -> (@out ()) to @escaping @callee_guaranteed (@guaranteed NIOPosix.NIOThread) -> () at <compiler-generated>, callback=0x00005555559a1ee0 swift-nioPackageTests.xctest`partial apply forwarder for closure #1 (NIOPosix.SelectableEventLoop) -> () in closure #1 (NIOPosix.NIOThread) -> () in static NIOPosix.MultiThreadedEventLoopGroup.setupThreadAndEventLoop(name: Swift.String, selectorFactory: () throws -> NIOPosix.Selector<NIOPosix.NIORegistration>, initializer: (NIOPosix.NIOThread) -> ()) -> NIOPosix.SelectableEventLoop at <compiler-generated>, self=NIOPosix.MultiThreadedEventLoopGroup) at MultiThreadedEventLoopGroup.swift:87:22
    frame #8: 0x000055555599cc8e swift-nioPackageTests.xctest`closure #1 in static MultiThreadedEventLoopGroup.setupThreadAndEventLoop(t=0x00007fffec000b60, selectorFactory=0x0000555555ef3320 swift-nioPackageTests.xctest`partial apply forwarder for closure #3 () throws -> NIOPosix.Selector<NIOPosix.NIORegistration> in NIOPosixTests.SALTest.setUpSAL() -> () at <compiler-generated>, initializer=0x00005555559a1d80 swift-nioPackageTests.xctest`partial apply forwarder for reabstraction thunk helper from @escaping @callee_guaranteed (@in_guaranteed NIOPosix.NIOThread) -> (@out ()) to @escaping @callee_guaranteed (@guaranteed NIOPosix.NIOThread) -> () at <compiler-generated>, lock=0x000055555690f240, _loop=0x00007fffec001930, loopUpAndRunningGroup=0x00005555568a4400) at MultiThreadedEventLoopGroup.swift:107:41
    frame #9: 0x00005555559a1e63 swift-nioPackageTests.xctest`partial apply for closure #1 in static MultiThreadedEventLoopGroup.setupThreadAndEventLoop(name:selectorFactory:initializer:) at <compiler-generated>:0
    frame #10: 0x000055555599b1ff swift-nioPackageTests.xctest`thunk for @escaping @callee_guaranteed (@guaranteed NIOThread) -> () at <compiler-generated>:0
    frame #11: 0x0000555555a0ac71 swift-nioPackageTests.xctest`partial apply for thunk for @escaping @callee_guaranteed (@guaranteed NIOThread) -> () at <compiler-generated>:0
    frame #12: 0x0000555555a0d7f2 swift-nioPackageTests.xctest`closure #1 in static ThreadOpsPosix.run($0=0x555556926600) at ThreadPosix.swift:105:13
    frame #13: 0x0000555555a0d959 swift-nioPackageTests.xctest`@objc closure #1 in static ThreadOpsPosix.run(handle:args:detachThread:) at <compiler-generated>:0
    frame #14: 0x00007ffff783a3f9 libpthread.so.0`start_thread + 233
    frame #15: 0x00007ffff6a534c3 libc.so.6`__clone + 67
```

**Comment by Lukasa** [comment]
That's interesting, this error is different: this is an assertion in the SelectableEventLoop, which is not what's showing up in the original report. That error doesn't really seem like it's possible: we've somehow exited the run loop, but failed to meet any of its exit conditions. I don't even really see how the syscall abstraction layer can trigger that failure.

**Comment by Lukasa** [comment]
Ok, I've continued to be unable to reproduce either of these failures. Can you be a bit more specific about your configuration? And do these failures persist if you throw away your `.build` directory and rebuild clean?

**Comment by Lukasa** [comment]
Oh, I see, the traceback is actually unrelated: we're hitting a _follow-on_ failure in the event loop thread caused by the crash on the main test thread. That's going to make things a little harder.

I suspect the reason you're seeing this is that you're hitting a timing window. It's pretty rare to hit a timing window that our own CI didn't hit! Given that things don't crash for a while, can we try to ascertain the relative progress of the various threads to work out what the ordering of events is? I'm afraid you'll need to do this, as I'm missing a reproducer.

Here's the current test code: https://github.com/apple/swift-nio/blob/06cb3dd0967f24a1e88dca4e2e06edf06aef8f3a/Tests/NIOPosixTests/SALEventLoopTests.swift#L35-L79

In this case there are three threads that are coordinating by way of condition variables and promises. The first is running on the main (test) thread: https://github.com/apple/swift-nio/blob/06cb3dd0967f24a1e88dca4e2e06edf06aef8f3a/Tests/NIOPosixTests/SALEventLoopTests.swift#L39-L45

The other two threads are in this block: https://github.com/apple/swift-nio/blob/06cb3dd0967f24a1e88dca4e2e06edf06aef8f3a/Tests/NIOPosixTests/SALEventLoopTests.swift#L47-L74

One of the two threads is the outer block, which runs on the event loop thread. The inner block is kicked off to a DispatchQueue and so is running on a thread of its own.

The intended behaviour of this test is to exit the event loop (causing it to be parked, which the main thread will wait for using `try self.assertParkedRightNow()`). Then, 100ms later we will enqueue a task onto the event loop that will block the loop for 500ms or until a semaphore is signaled, whichever is first (this wakes up the loop, which the main thread waits for using `try self.assertWakeup()`). After that task is enqueued, we will then enqueue 11 more tasks, signal the semaphore, and then use the result of the last task to signal the end of the test. We expect all 11 of these tasks to complete without needing to wake the loop up again, so the entire test should complete with only one further loop park and no further wakeups.

I think a very useful thing to do in the first instance would be to add `print` statements essentially around each line and see what output we get. Assuming the test continues to fail in the same way, we can use that to plot an ordering of events that leads to this hang. (We should remove the change of `try` to `try!`.)

**Comment by buttaface** [comment]
Alright, I'll experiment with it a bit and let you know tomorrow.

**Comment by buttaface** [comment]
Almost two months later, I finally got around to this. I added a bunch of print statements, as you said:
```
diff --git a/Tests/NIOPosixTests/SALEventLoopTests.swift b/Tests/NIOPosixTests/SALEventLoopTests.swift
index 56f653d3..46b70571 100644
--- a/Tests/NIOPosixTests/SALEventLoopTests.swift
+++ b/Tests/NIOPosixTests/SALEventLoopTests.swift
@@ -51,6 +51,7 @@ final class SALEventLoopTests: XCTestCase, SALTest {
             DispatchQueue(label: "background").asyncAfter(deadline: .now() + .milliseconds(100)) {
                 let semaphore = DispatchSemaphore(value: 0)
 
+print("before wedging")
                 thisLoop.execute {
                     // Wedge the loop open. This will also _wake_ the loop.
                     XCTAssertEqual(semaphore.wait(timeout: .now() + .milliseconds(500)), .success)
@@ -60,21 +61,29 @@ final class SALEventLoopTests: XCTestCase, SALTest {
                 // Now execute 10 tasks.
                 var i = 0
                 for _ in 0..<10 {
+print("execute loop \(i)")
                     thisLoop.execute {
+print("executing \(i)")
                         i &+= 1
                     }
                 }
 
+print("before last \(i)")
                 // Now enqueue a "last" task.
                 let lastTask = thisLoop.submit { i &+= 1 }
 
+print("before unblocking")
                 // Now we can unblock the semaphore.
                 semaphore.signal()
 
+print("before cascade \(i)")
                 lastTask.cascade(to: promise)
+print("after cascade \(i)")
             }
             
+print("before return")
             return promise.futureResult
         }.salWait()
+print("finishing up")
     }
 }
```
With that added, I get the following passing output when just running `swift test`:
```
Test Case 'SALEventLoopTests.testSchedulingTaskOnSleepingLoopWakesUpOnce' started at 2021-11-19 19:01:30.102
before return
before wedging
execute loop 0
execute loop 0
execute loop 0
execute loop 0
execute loop 0
execute loop 0
execute loop 0
execute loop 0
execute loop 0
execute loop 0
before last 0
before unblocking
before cascade 0
after cascade 0
Unblocking wedged task
executing 0
executing 1
executing 2
executing 3
executing 4
executing 5
executing 6
executing 7
executing 8
executing 9
finishing up
Test Case 'SALEventLoopTests.testSchedulingTaskOnSleepingLoopWakesUpOnce' passed (0.101 seconds)
```
If I add the `--parallel` or `--filter` flags, it fails like this:
```
Test Case 'SALEventLoopTests.testSchedulingTaskOnSleepingLoopWakesUpOnce' started at 2021-11-19 19:05:07.736
before return
before wedging
execute loop 0
execute loop 0
execute loop 0
execute loop 0
execute loop 0
execute loop 0
execute loop 0
execute loop 0
execute loop 0
execute loop 0
before last 0
before unblocking
Unblocking wedged task
executing 0
executing 1
executing 2
executing 3
executing 4
executing 5
executing 6
executing 7
executing 8
executing 9
before cascade 11
after cascade 11
<EXPR>:0: error: SALEventLoopTests.testSchedulingTaskOnSleepingLoopWakesUpOnce : threw error "TimeoutError(description: "/home/butta/swift-nio/Tests/NIOPosixTests/SyscallAbstractionLayer.swift:460")"
Test Case 'SALEventLoopTests.testSchedulingTaskOnSleepingLoopWakesUpOnce' failed (5.103 seconds)
```
The only difference is the order and iteration value of those last two cascade statements, which seems to break something and it never finishes up properly.

Let me know what you think, this was with the system Swift 5.5.1 package in Fedora Core 33 x86_64 running on the same single-core VPS.

**Comment by Lukasa** [comment]
Yup, this looks like the error to me.

The load bearing events are "Unblocking wedged task" and "before cascade"/"after cascade". In the first case, we apply the cascade _before_ the unblock occurs. In the second case, we apply it after (and, in fact, we end up applying it after the queue of tasks has executed).

The reason this matters is the definition of `cascade(to:)`:

https://github.com/apple/swift-nio/blob/addf69cfe60376c325397c8926589415576b1dd1/Sources/NIOCore/EventLoopFuture.swift#L865-L873

This is ultimately just a wrapper around `whenComplete`, itself a wrapper around the private method `_whenComplete`:

https://github.com/apple/swift-nio/blob/addf69cfe60376c325397c8926589415576b1dd1/Sources/NIOCore/EventLoopFuture.swift#L703-L711

Importantly, this method has to acquire event loop context when it is called from off the event loop. In this case, that has to happen: we're on a private `DispatchQueue`. To acquire the event loop context we'll need to wake the loop up, but that can't happen until the SAL waits for it. This test is attempting to assert that we _will not_ wake the loop again, so naturally the loop fails to wake.

The fix for this is simply to complete the promise within the `lastTask` block, and then remove the `cascade` (and so the temporary `lastTask` variable. This should entirely resolve the problem.

@buttaface do you feel like prototyping this fix? If it works for you, we'll happily take a PR that applies it.

**Comment by buttaface** [comment]
I'm not familiar with this code, so if you paste the patch you have in mind, I can apply it and see if it works.

Why does this only break in a single-core environment with those two flags?

**Comment by Lukasa** [comment]
> I'm not familiar with this code, so if you paste the patch you have in mind, I can apply it and see if it works.

```diff
diff --git a/Tests/NIOPosixTests/SALEventLoopTests.swift b/Tests/NIOPosixTests/SALEventLoopTests.swift
index 56f653d3..8f6df2db 100644
--- a/Tests/NIOPosixTests/SALEventLoopTests.swift
+++ b/Tests/NIOPosixTests/SALEventLoopTests.swift
@@ -66,12 +66,13 @@ final class SALEventLoopTests: XCTestCase, SALTest {
                 }
 
                 // Now enqueue a "last" task.
-                let lastTask = thisLoop.submit { i &+= 1 }
+                thisLoop.execute {
+                    i &+= 1
+                    promise.succeed(())
+                }
 
                 // Now we can unblock the semaphore.
                 semaphore.signal()
-
-                lastTask.cascade(to: promise)
             }
             
             return promise.futureResult
```

> Why does this only break in a single-core environment with those two flags?

I don't think it _only_ breaks in this environment, I think it's _much more likely_ to break in that environment. The issue here is simply a timing one. It has always been possible for the thread running in the background `DispatchQueue` to be take longer between calling `semaphore.signal` and `promise.cascade(to:)` than it does for the event loop to process the tasks. If that happens, this test will fail in this way.

Your environment makes this much more likely. In the case of `--parallel`, you're increasing the load on the system. This makes it more likely that our thread will suspend between calling `signal` and `cascade(to:)`. If it does, the window for this error widens substantially. In the case of `--filter` you're still slowing the system down for _this_ test, because all statics are being initialised in this test code. In either case I think it becomes much more likely. A dual-core system can run both threads in parallel so there is less chance of the queue thread to be suspended, and so it's probably going to win the race to enqueue a task before the loop sleeps.

**Comment by buttaface** [comment]
Yep, that fixes it.
